### PR TITLE
Use evented dataclass for camera

### PR DESCRIPTION
**Pull request by sofroniewn**
# Description
Following on from #1910, #1911, #1912, and #1913 this PR switches to using an evented dataclass for the camera. Unlike the others this one is a little more complicated as the camera has a private copy of the dims that allows it to do some special checking the length of `ndisplay` property and the center of the camera match. I'm not sure this is really necessary or best done here - i.e. another object which knows about both camera and dims could be responsible for checking that, and so we could simplify this dataclass to make it more in line with the others ....

## Type of change

- [x] Refactor (non-breaking change which fixes an issue)

### CODE REVIEW

**Comment by sofroniewn** [comment]
I've updated this to drop `dims` from `camera` which I think makes everything much cleaner. We can decide if we like this when the PR is ready for review, which might not be for a bit

**Comment by sofroniewn** [comment]
Note that by dropping the `dims` from `camera` we've also had to drop `viewer.camera.ndisplay`, instead people have to use `viewer.dims.ndisplay` we could deprecate/ add error message

**Comment by sofroniewn** [comment]
I've now updated from master. This PR drops the `viewer.camera.ndisplay` method which was redundant with `viewer.dims.ndisplay` as it wasn't necessary to have the duplicated property

**Comment by Czaki** [comment]
For synchronization of two viewers, I currently use vispy camera object `viewer_widget.view.camera.get_state()` It will be nice to shortcut this to `self.viewer.camera.get_state()` and `self.viewer.camera.set_state()`. 

As I look there should be no problem with adding such methods to dataclass camera, but I would like to ask if I'm correct?

**Comment by sofroniewn** [comment]
> For synchronization of two viewers, I currently use vispy camera object viewer_widget.view.camera.get_state() It will be nice to shortcut this to self.viewer.camera.get_state() and self.viewer.camera.set_state().
>
> As I look there should be no problem with adding such methods to dataclass camera, but I would like to ask if I'm correct?

This will be a very common need for all our evented dataclasses and how I imagine we implement things like #851 which we'll use frequently (like in making movies in #780). I had thought on the get side we'd use

```python
from dataclasses import asdict

asdict(viewer.camera)
```

and on the set side we'd have to loop through a dictionary and update them. I'm not sure if there is an easier way.

Question is if we want convenience methods for these living on our `evented_dataclass` directly, and if so what should they be called? It moves the API further away from being a drop in replacement for a dataclass, but it is a convenience that we'll want to re-use. Curious how @tlambert03 and @jni think about this? But I definitely don't think it is something that should live in this PR as its generic functionality that's not camera specific.

One option would be to provide a `fromdict` method that took a dataclass and a dict and did an inplace update of the dataclass. Then we'd have `asdict` and `fromdict`. Question is should we allow then to live on the dataclass itself?

**Comment by tlambert03** [comment]
> Question is if we want convenience methods for these living on our evented_dataclass directly, and if so what should they be called?

the `asdict` convenience is already on the `evented_dataclass` here:
https://github.com/napari/napari/blob/6f28b00fcd3353c16496e892cf92f2586e766fe8/napari/utils/events/dataclass.py#L480
(I don't particularly love "`_get_state`"... that's just what we'd been using, so I went with it ... happy to rename it, and honeslty "asdict" seems like as good of a method name as any)

**Comment by sofroniewn** [comment]
> (I don't particularly love "_get_state"... that's just what we'd been using, so I went with it ... happy to rename it, and honeslty "asdict" seems like as good of a method name as any)

Yeah, we could rename `_get_state` as `asdict` and add a `fromdict` method too. I'd be open to a new PR that did that @Czaki. We'd then phase out and eliminate the "_get_state" methods.

**Comment by tlambert03** [comment]
> Yeah, we could rename _get_state as asdict and add a fromdict method too. I'd be open to a new PR that did that @Czaki. We'd then phase out and eliminate the "_get_state" methods.

+1
(with `fromdict` as a `@classmethod`, yes?)

**Comment by sofroniewn** [comment]
> (with fromdict as a @classmethod, yes?)

very nice

File: napari/components/camera.py [hunk 1]
```diff
@@ -1,102 +1,23 @@
...
-        else:
-            return (0, 0, 90)
-
-    @angles.setter
-    def angles(self, angles):
-        if self._angles == tuple(angles):
-            return
-        self._angles = tuple(angles)
-        self.events.angles()
+    center: Property[Tuple, None, tuple] = (0, 0, 0)
```
Comment (tlambert03) [inline napari/components/camera.py]:
```suggestion
    center: Property[Tuple[float, float, float], None, tuple] = (0, 0, 0)
```

right? 
You do seem to be removing the `Center must be same length as currently displayed` check here... but if you want to allow for any length tuple of ints, then use `Tuple[float, ...]` (though actually... I was gonna suggest that maybe we consider standardizing the length of the camera center to always being 3, even in 2D...)

same Tuple[type] thing for angles...

also, zoom is `float` in the docs but `int` here... is that correct?

**Review by tlambert03 — APPROVED** [review]
looks good to me (outside of a couple type clarifications), assuming you're ok with the couple conditional checks you've lost here during the conversion.

I like the removal of `dims`

File: napari/components/camera.py (hunk 1 above)
Comment (sofroniewn) [reply napari/components/camera.py]: > I was gonna suggest that maybe we consider standardizing the length of the camera center to always being 3, even in 2D.

I like this idea! If someone passes a 2-tuple we should pad to a 3-tuple

> same Tuple[type] thing for angles...

will do

> also, zoom is float in the docs but int here... is that correct?

should be float everywhere

Will make these changes soon!

**Comment by sofroniewn** [comment]
@tlambert03 I've tried to update as you suggested, can you take one more look though if this is what you had in mind?

I've also added `Len_3_Tuple = Tuple[float, float, float]` as otherwise the type part became really long and multiline in the dataclass, but not sure if this will mess up typing.

Curious if @jni wants to quickly weigh in before merge

File: napari/utils/misc.py [hunk 2]
```diff
@@ -336,3 +336,24 @@ def all_subclasses(cls: Type) -> set:
     return set(cls.__subclasses__()).union(
         [s for c in cls.__subclasses__() for s in all_subclasses(c)]
     )
+
+
+def force_3_tuple(value):
```
Comment (tlambert03) [inline napari/utils/misc.py]: A bit more reusable:
```python

def force_n_tuple(val, n, fill=0):
    assert n > 0, 'n must be greater than 0'
    tuple_value = tuple(val)
    return (fill,) * (n - len(tuple_value)) + tuple_value[-n:]

force_3_tuple = partial(force_n_tuple, n=3)
```

(And maybe "ensure_3_tuple" to be consistent with similar functions elsewhere?)

**Review by tlambert03 — COMMENTED** [review]
Yep looks good!  I might have a _slight_ preference for using (0, 0, 0) for the default as you used to have it (instead of (0,) * 3 ...) but typing all looks good and the type alias is fine.

File: napari/utils/misc.py (hunk 2 above)
Comment (sofroniewn) [reply napari/utils/misc.py]: done!

**Comment by sofroniewn** [comment]
I've made the requested changes. Thanks for the review @tlambert03 will merge now!

**Comment by Czaki** [comment]
> > (I don't particularly love "_get_state"... that's just what we'd been using, so I went with it ... happy to rename it, and honeslty "asdict" seems like as good of a method name as any)
> 
> Yeah, we could rename `_get_state` as `asdict` and add a `fromdict` method too. I'd be open to a new PR that did that @Czaki. We'd then phase out and eliminate the "_get_state" methods.

I will do PD, but I do not like idea of iteration over field to set it values in user code. It is so common that there should be method for this. 

And `fromdict` as dataclss is eqivalent of `Constructor(**dict)`

**Pull request merged**